File: prayer_app_expo/src/screens/QiblaScreen.js
import React, { useState, useEffect, useRef, useCallback } from 'react';
import { View, Text, ScrollView, StyleSheet, Animated } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import Svg, { Circle, Line, Polygon, Rect, Text as SvgText } from 'react-native-svg';
import * as Location from 'expo-location';
import { Spacing, QiblaLayout, interFont, getTypography } from '../theme/theme';
import { useTheme } from '../providers/ThemeProvider';
import { useLocation } from '../providers/LocationProvider';
import { computeQiblaDegrees } from '../services/qiblaService';

const SMOOTH_ALPHA = 0.4;
const ANIM_DURATION = 50;
const TEXT_THROTTLE_MS = 200; // only re-render for text updates this often

// Arrow points UP at 0° rotation — no baseline correction needed
const ARROW_BASELINE_DEG = 0;

// ── Angle helpers ──
function normalizeAngle(a) {
    return ((a % 360) + 360) % 360;
}
function shortestDiff(from, to) {
    let diff = normalizeAngle(to - from);
    if (diff > 180) diff -= 360;
    return diff; // -180..180
}

export default function QiblaScreen() {
    const { theme: tc } = useTheme();
    const { location } = useLocation();
    const typo = getTypography(tc);

    const loc = location || {
        city: 'Bucharest', country: 'Romania',
        lat: 44.4268, lon: 26.1025, source: 'default',
    };
    const cityLabel = loc.country ? `${loc.city}, ${loc.country}` : loc.city;
    const degrees = computeQiblaDegrees(loc.lat, loc.lon);

    // heading state is only used for direction text — throttled, not per-update
    const [heading, setHeading] = useState(null);
    const [compassAvailable, setCompassAvailable] = useState(true);

    // Refs for high-frequency sensor path (no re-renders)
    const smoothedRef = useRef(null);
    const animatedRotation = useRef(new Animated.Value(0)).current;
    const cumulativeRotRef = useRef(0);
    const prevSmoothedRef = useRef(null);
    const lastTextUpdateRef = useRef(0);

    // Update frequency tracking (first 5 seconds)
    const freqStartRef = useRef(null);
    const freqCountRef = useRef(0);
    const freqLoggedRef = useRef(false);

    useEffect(() => {
        let sub;
        (async () => {
            const { status } = await Location.requestForegroundPermissionsAsync();
            if (status !== 'granted') {
                setCompassAvailable(false);
                console.log('[Qibla] Location permission denied');
                return;
            }

            sub = await Location.watchHeadingAsync((headingData) => {
                const rawHeading = (headingData.trueHeading >= 0)
                    ? headingData.trueHeading
                    : headingData.magHeading;
                if (rawHeading < 0) return;

                const now = Date.now();

                // ── Update frequency tracking (first 5s) ──
                if (freqStartRef.current === null) freqStartRef.current = now;
                freqCountRef.current++;
                const elapsed = now - freqStartRef.current;
                if (elapsed >= 5000 && !freqLoggedRef.current) {
                    const hz = (freqCountRef.current / (elapsed / 1000)).toFixed(1);
                    console.log(`[Qibla] Update rate: ${hz} Hz (${freqCountRef.current} updates in ${(elapsed / 1000).toFixed(1)}s)`);
                    freqLoggedRef.current = true;
                }

                // ── Smoothing (high-frequency, no setState) ──
                if (smoothedRef.current === null) {
                    smoothedRef.current = rawHeading;
                } else {
                    const diff = shortestDiff(smoothedRef.current, rawHeading);
                    smoothedRef.current = normalizeAngle(smoothedRef.current + diff * SMOOTH_ALPHA);
                }

                const smoothed = smoothedRef.current;

                // ── Cumulative rotation → Animated.Value (no re-render) ──
                if (prevSmoothedRef.current === null) {
                    cumulativeRotRef.current = -smoothed;
                } else {
                    const rotDiff = shortestDiff(prevSmoothedRef.current, smoothed);
                    cumulativeRotRef.current -= rotDiff;
                }
                prevSmoothedRef.current = smoothed;

                Animated.timing(animatedRotation, {
                    toValue: cumulativeRotRef.current,
                    duration: ANIM_DURATION,
                    useNativeDriver: true,
                }).start();

                // ── Throttled setState for text/debug only ──
                if (now - lastTextUpdateRef.current >= TEXT_THROTTLE_MS) {
                    lastTextUpdateRef.current = now;
                    setHeading(Math.round(smoothed * 10) / 10);
                }
            });
            console.log('[Qibla] Location heading subscription started');
        })();
        return () => { if (sub) sub.remove(); };
    }, []);

    // Direction guidance (only re-computed on throttled heading updates)
    let statusText = '';
    let facingQibla = false;
    if (compassAvailable && heading !== null) {
        const diff = shortestDiff(heading, degrees);
        if (Math.abs(diff) < 5) {
            statusText = 'Facing Qibla ✓';
            facingQibla = true;
        } else if (diff > 0) {
            statusText = 'Turn right to face Qibla';
        } else {
            statusText = 'Turn left to face Qibla';
        }
    }

    const hasHeading = compassAvailable && heading !== null;
    const animatedStyle = {
        transform: [{
            rotate: animatedRotation.interpolate({
                inputRange: [-3600, 3600],
                outputRange: ['-3600deg', '3600deg'],
            })
        }],
    };

    // Debug delta for overlay
    const debugDelta = hasHeading ? normalizeAngle(degrees - heading).toFixed(1) : '--';

    return (
        <ScrollView style={styles.scroll} contentContainerStyle={styles.content}>
            <View style={{ height: QiblaLayout.titleMarginTop }} />

            {/* Title */}
            <View style={styles.padH}>
                <Text style={typo.titleMedium}>Qibla</Text>
            </View>
            <View style={{ height: Spacing.s16 }} />

            {/* City row */}
            <View style={styles.cityRow}>
                <MaterialCommunityIcons name="map-marker" size={QiblaLayout.cityIconSize} color={tc.accent} />
                <View style={{ width: 4 }} />
                <Text style={{
                    fontFamily: interFont('500'),
                    fontSize: QiblaLayout.cityFontSize,
                    color: tc.textMuted,
                }}>{cityLabel}</Text>
            </View>
            <View style={{ height: Spacing.s24 }} />

            {/* Big degree */}
            <Text style={[styles.degreeText, { color: tc.accent }]}>{degrees.toFixed(1)}°</Text>
            <Text style={[styles.degreeSubtitle, { color: tc.textMuted }]}>from North</Text>
            <View style={{ height: Spacing.s24 }} />

            {/* Compass dial */}
            <View style={styles.compassBox}>
                {/* Fixed pointer triangle at top */}
                {hasHeading && (
                    <View style={styles.pointerContainer}>
                        <Svg width={QiblaLayout.pointerSize} height={QiblaLayout.pointerSize}>
                            <Polygon
                                points={`${QiblaLayout.pointerSize / 2},${QiblaLayout.pointerSize} 0,0 ${QiblaLayout.pointerSize},0`}
                                fill={tc.accent}
                            />
                        </Svg>
                    </View>
                )}
                <Animated.View style={hasHeading ? animatedStyle : undefined}>
                    <CompassDial degrees={degrees} tc={tc} showPointer={!hasHeading} />
                </Animated.View>
            </View>

            {/* Debug overlay (temporary) */}
            {hasHeading && (
                <View style={[styles.debugOverlay, { backgroundColor: tc.card, borderColor: tc.cardBorder }]}>
                    <Text style={{ fontFamily: interFont('400'), fontSize: 11, color: tc.textMuted }}>
                        heading={heading}° | qibla={degrees}° | delta={debugDelta}°
                    </Text>
                </View>
            )}

            {/* Compass not available notice */}
            {!compassAvailable && (
                <>
                    <View style={{ height: Spacing.s8 }} />
                    <View style={styles.padH}>
                        <View style={[styles.unavailableCard, {
                            backgroundColor: tc.card,
                            borderColor: tc.cardBorder,
                        }]}>
                            <MaterialCommunityIcons name="compass-off" size={16} color={tc.textMuted} />
                            <View style={{ width: 8 }} />
                            <Text style={{
                                fontFamily: interFont('500'),
                                fontSize: 12,
                                color: tc.textMuted,
                            }}>Compass not available on this device.</Text>
                        </View>
                    </View>
                </>
            )}
            <View style={{ height: Spacing.s24 }} />

            {/* Status text */}
            <View style={styles.padH}>
                {facingQibla ? (
                    <Text style={[styles.statusText, { color: tc.accent, fontFamily: interFont('600') }]}>
                        {statusText}
                    </Text>
                ) : hasHeading ? (
                    <Text style={[styles.statusText, { color: tc.textMuted }]}>
                        {statusText}
                    </Text>
                ) : (
                    <Text style={[styles.statusText, { color: tc.textMuted }]}>
                        Qibla is at{' '}
                        <Text style={{ color: tc.accent, fontFamily: interFont('600') }}>
                            {degrees.toFixed(1)}°
                        </Text>
                        {' '}from North.
                    </Text>
                )}
            </View>
            <View style={{ height: Spacing.s24 }} />

            {/* Kaaba label card */}
            <View style={styles.padH}>
                <View style={[styles.kaabaCard, {
                    backgroundColor: tc.card,
                    borderColor: tc.cardBorder,
                }]}>
                    <Text style={[styles.arabicText, { color: tc.accent }]}>
                        {'\u0627\u0644\u0643\u0639\u0628\u0629 \u0627\u0644\u0645\u0634\u0631\u0651\u0641\u0629'}
                    </Text>
                    <View style={{ height: 4 }} />
                    <Text style={[styles.translitText, { color: tc.textMuted }]}>
                        Al-Kaaba Al-Musharrafah
                    </Text>
                </View>
            </View>
            <View style={{ height: QiblaLayout.screenPadding }} />
        </ScrollView>
    );
}

// ────────────────────────────────────────────────
// COMPASS DIAL (SVG) — memoized to avoid redraws
// ────────────────────────────────────────────────
const CompassDial = React.memo(function CompassDial({ degrees, tc, showPointer }) {
    const size = QiblaLayout.compassSize;
    const cx = size / 2;
    const cy = size / 2;
    const radius = size / 2 - 20;

    const ticks = [];
    for (let deg = 0; deg < 360; deg += 5) {
        const isMajor = deg % 90 === 0;
        const is30 = deg % 30 === 0;
        const len = isMajor
            ? QiblaLayout.tickLengthMajor
            : (is30 ? QiblaLayout.tickLength + 2 : QiblaLayout.tickLength);
        const rad = (deg - 90) * Math.PI / 180;
        const x1 = cx + radius * Math.cos(rad);
        const y1 = cy + radius * Math.sin(rad);
        const x2 = cx + (radius - len) * Math.cos(rad);
        const y2 = cy + (radius - len) * Math.sin(rad);
        ticks.push(
            <Line
                key={`t${deg}`}
                x1={x1} y1={y1} x2={x2} y2={y2}
                stroke={isMajor || is30 ? tc.textMuted + 'CC' : tc.textMuted + '80'}
                strokeWidth={isMajor || is30 ? 1.5 : 1}
            />
        );
    }

    // Cardinal labels
    const cardinals = [
        { label: 'N', deg: 0 },
        { label: 'E', deg: 90 },
        { label: 'S', deg: 180 },
        { label: 'W', deg: 270 },
    ];
    const labels = cardinals.map(({ label, deg }) => {
        const rad = (deg - 90) * Math.PI / 180;
        const labelR = radius - QiblaLayout.tickLengthMajor - 12;
        const x = cx + labelR * Math.cos(rad);
        const y = cy + labelR * Math.sin(rad);
        const isN = label === 'N';
        return (
            <SvgText
                key={label}
                x={x} y={y + 5}
                textAnchor="middle"
                fontSize={QiblaLayout.cardinalFontSize}
                fontWeight="700"
                fill={isN ? tc.accent : tc.textMuted + '99'}
            >
                {label}
            </SvgText>
        );
    });

    // Pointer triangle at top (only in static mode)
    const pointerY = cy - radius - 4;
    const ps = QiblaLayout.pointerSize;
    const pointerPoints = `${cx},${pointerY} ${cx - ps / 2},${pointerY - ps} ${cx + ps / 2},${pointerY - ps}`;

    // Qibla needle
    const needleRad = (degrees - 90) * Math.PI / 180;
    const needleLen = radius - QiblaLayout.tickLengthMajor - 24;
    const nx = cx + needleLen * Math.cos(needleRad);
    const ny = cy + needleLen * Math.sin(needleRad);

    // Kaaba marker on ring
    const kx = cx + (radius + 2) * Math.cos(needleRad);
    const ky = cy + (radius + 2) * Math.sin(needleRad);
    const ksq = QiblaLayout.kaabaIconSize * 0.55;

    return (
        <Svg width={size} height={size} viewBox={`0 0 ${size} ${size}`}>
            {/* Outer ring */}
            <Circle
                cx={cx} cy={cy} r={radius}
                stroke={tc.cardBorder} strokeWidth={QiblaLayout.compassStroke}
                fill="none"
            />
            {/* Ticks */}
            {ticks}
            {/* Labels */}
            {labels}
            {/* Pointer triangle (only static mode) */}
            {showPointer && <Polygon points={pointerPoints} fill={tc.accent} />}
            {/* Needle */}
            <Line
                x1={cx} y1={cy} x2={nx} y2={ny}
                stroke={tc.accent} strokeWidth={QiblaLayout.needleWidth}
                strokeLinecap="round"
            />
            {/* Center dot */}
            <Circle cx={cx} cy={cy} r={QiblaLayout.centerDotRadius} fill={tc.accent} />
            {/* Kaaba marker */}
            <Circle cx={kx} cy={ky} r={QiblaLayout.kaabaIconSize / 2 + 3} fill={tc.accent} />
            <Rect
                x={kx - ksq / 2} y={ky - ksq / 2}
                width={ksq} height={ksq}
                fill="#000"
            />
        </Svg>
    );
});

// ────────────────────────────────────────────────
// STYLES
// ────────────────────────────────────────────────
const styles = StyleSheet.create({
    scroll: { flex: 1 },
    content: { paddingBottom: 0 },
    padH: { paddingHorizontal: QiblaLayout.screenPadding },
    cityRow: {
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
    },
    degreeText: {
        fontFamily: interFont('700'),
        fontSize: QiblaLayout.degreeFontSize,
        textAlign: 'center',
        lineHeight: QiblaLayout.degreeFontSize * 1.1,
    },
    degreeSubtitle: {
        fontFamily: interFont('500'),
        fontSize: QiblaLayout.degreeSubtitleSize,
        textAlign: 'center',
    },
    compassBox: {
        alignItems: 'center',
        justifyContent: 'center',
    },
    pointerContainer: {
        position: 'absolute',
        top: 20 - QiblaLayout.pointerSize - 4,
        zIndex: 10,
    },
    debugOverlay: {
        marginTop: 8,
        marginHorizontal: QiblaLayout.screenPadding,
        paddingHorizontal: 12,
        paddingVertical: 6,
        borderRadius: 8,
        borderWidth: 1,
        alignItems: 'center',
    },
    statusText: {
        fontFamily: interFont('400'),
        fontSize: QiblaLayout.statusFontSize,
        textAlign: 'center',
    },
    unavailableCard: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'center',
        paddingHorizontal: 16,
        paddingVertical: 10,
        borderRadius: 12,
        borderWidth: 1,
    },
    kaabaCard: {
        padding: QiblaLayout.cardPadding,
        borderRadius: QiblaLayout.cardRadius,
        borderWidth: 1,
        alignItems: 'center',
    },
    arabicText: {
        fontFamily: interFont('600'),
        fontSize: QiblaLayout.arabicFontSize,
        textAlign: 'center',
    },
    translitText: {
        fontFamily: interFont('500'),
        fontSize: QiblaLayout.translitFontSize,
        textAlign: 'center',
    },
});

File: prayer_app_flutter/lib/src/screens/qibla_screen.dart
import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'package:material_design_icons_flutter/material_design_icons_flutter.dart';
import 'package:flutter_compass/flutter_compass.dart';
import '../theme/app_theme.dart';
import '../providers/theme_provider.dart';
import '../services/location_service.dart';
import '../services/qibla_service.dart';

class QiblaScreen extends StatefulWidget {
  final LocationNotifier locationNotifier;

  const QiblaScreen({super.key, required this.locationNotifier});

  @override
  State<QiblaScreen> createState() => _QiblaScreenState();
}

class _QiblaScreenState extends State<QiblaScreen> {
  StreamSubscription<CompassEvent>? _compassSub;
  double? _heading; // smoothed heading in degrees 0..360
  bool _compassAvailable = true;

  static const double _smoothAlpha = 0.2;

  @override
  void initState() {
    super.initState();
    _initCompass();
  }

  Future<void> _initCompass() async {
    // flutter_compass exposes a nullable stream — null means unavailable
    final stream = FlutterCompass.events;
    if (stream == null) {
      if (mounted) setState(() => _compassAvailable = false);
      debugPrint('[Qibla] Compass not available (stream null)');
      return;
    }

    _compassSub = stream.listen(
      (event) {
        final raw = event.heading;
        if (raw == null) return;
        final normalized = (raw + 360) % 360;
        setState(() {
          if (_heading == null) {
            _heading = normalized;
          } else {
            // Low-pass filter with circular interpolation
            var diff = normalized - _heading!;
            if (diff > 180) diff -= 360;
            if (diff < -180) diff += 360;
            _heading = (_heading! + _smoothAlpha * diff + 360) % 360;
          }
        });
        debugPrint('[Qibla] heading=${_heading!.toStringAsFixed(1)}');
      },
      onError: (_) {
        if (mounted) setState(() => _compassAvailable = false);
        debugPrint('[Qibla] Compass error — marking unavailable');
      },
    );
  }

  @override
  void dispose() {
    _compassSub?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final tc = ThemeScope.of(context).current;

    return ListenableBuilder(
      listenable: widget.locationNotifier,
      builder: (context, _) {
        final loc = widget.locationNotifier.data;
        final qiblaBearing = computeQiblaDegrees(loc.lat, loc.lon);
        final cityLabel = loc.country.isNotEmpty
            ? '${loc.city}, ${loc.country}'
            : loc.city;

        // Compute direction guidance
        String statusText;
        bool facingQibla = false;
        if (_compassAvailable && _heading != null) {
          var diff = qiblaBearing - _heading!;
          if (diff > 180) diff -= 360;
          if (diff < -180) diff += 360;
          if (diff.abs() < 5) {
            statusText = 'Facing Qibla ✓';
            facingQibla = true;
          } else if (diff > 0) {
            statusText = 'Turn right to face Qibla';
          } else {
            statusText = 'Turn left to face Qibla';
          }
        } else {
          statusText = 'Qibla is at ${qiblaBearing.toStringAsFixed(1)}° from North.';
        }

        return ListView(
          padding: EdgeInsets.zero,
          children: [
            const SizedBox(height: QiblaLayout.titleMarginTop),

            // ── Title ──
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: QiblaLayout.screenPadding),
              child: Text('Qibla', style: AppTypography.titleMedium(tc)),
            ),
            const SizedBox(height: AppSpacing.s16),

            // ── City row ──
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Icon(MdiIcons.mapMarker, color: tc.accent, size: QiblaLayout.cityIconSize),
                const SizedBox(width: 4),
                Text(
                  cityLabel,
                  style: GoogleFonts.inter(
                    fontSize: QiblaLayout.cityFontSize,
                    color: tc.textMuted,
                    fontWeight: FontWeight.w500,
                  ),
                ),
              ],
            ),
            const SizedBox(height: AppSpacing.s24),

            // ── Big degree ──
            Center(
              child: Text(
                '${qiblaBearing.toStringAsFixed(1)}°',
                style: GoogleFonts.inter(
                  fontSize: QiblaLayout.degreeFontSize,
                  fontWeight: FontWeight.w700,
                  color: tc.accent,
                  height: 1.1,
                ),
              ),
            ),
            Center(
              child: Text(
                'from North',
                style: GoogleFonts.inter(
                  fontSize: QiblaLayout.degreeSubtitleSize,
                  color: tc.textMuted,
                  fontWeight: FontWeight.w500,
                ),
              ),
            ),
            const SizedBox(height: AppSpacing.s24),

            // ── Compass dial or unavailable ──
            if (_compassAvailable && _heading != null) ...[
              Center(
                child: SizedBox(
                  width: QiblaLayout.compassSize,
                  height: QiblaLayout.compassSize,
                  child: Stack(
                    children: [
                      // Pointer triangle stays fixed at top
                      Positioned(
                        left: QiblaLayout.compassSize / 2 - QiblaLayout.pointerSize / 2,
                        top: 20 - QiblaLayout.pointerSize - 4,
                        child: CustomPaint(
                          size: Size(QiblaLayout.pointerSize, QiblaLayout.pointerSize),
                          painter: _PointerPainter(color: tc.accent),
                        ),
                      ),
                      // Rotating compass + needle
                      Transform.rotate(
                        angle: -_heading! * pi / 180,
                        child: CustomPaint(
                          size: const Size(QiblaLayout.compassSize, QiblaLayout.compassSize),
                          painter: _CompassPainter(
                            qiblaDegrees: qiblaBearing,
                            accent: tc.accent,
                            muted: tc.textMuted,
                            ring: tc.cardBorder,
                            needle: tc.accent,
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
              ),
            ] else ...[
              // static compass (no heading available)
              Center(
                child: SizedBox(
                  width: QiblaLayout.compassSize,
                  height: QiblaLayout.compassSize,
                  child: CustomPaint(
                    painter: _CompassPainter(
                      qiblaDegrees: qiblaBearing,
                      accent: tc.accent,
                      muted: tc.textMuted,
                      ring: tc.cardBorder,
                      needle: tc.accent,
                    ),
                  ),
                ),
              ),
              const SizedBox(height: AppSpacing.s8),
              // Unavailable notice
              if (!_compassAvailable)
                Padding(
                  padding: const EdgeInsets.symmetric(horizontal: QiblaLayout.screenPadding),
                  child: Container(
                    padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 10),
                    decoration: BoxDecoration(
                      color: tc.card,
                      borderRadius: BorderRadius.circular(12),
                      border: Border.all(color: tc.cardBorder),
                    ),
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Icon(MdiIcons.compassOff, size: 16, color: tc.textMuted),
                        const SizedBox(width: 8),
                        Text(
                          'Compass not available on this device.',
                          style: GoogleFonts.inter(
                            fontSize: 12,
                            color: tc.textMuted,
                            fontWeight: FontWeight.w500,
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
            ],
            const SizedBox(height: AppSpacing.s8),

            // ── Debug overlay (temporary) ──
            if (_compassAvailable && _heading != null)
              Padding(
                padding: const EdgeInsets.symmetric(horizontal: QiblaLayout.screenPadding),
                child: Container(
                  padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
                  decoration: BoxDecoration(
                    color: tc.card,
                    borderRadius: BorderRadius.circular(8),
                    border: Border.all(color: tc.cardBorder),
                  ),
                  child: Center(
                    child: Text(
                      'heading=${_heading!.toStringAsFixed(1)}° | qibla=${qiblaBearing.toStringAsFixed(1)}° | delta=${((qiblaBearing - _heading! + 360) % 360).toStringAsFixed(1)}°',
                      style: GoogleFonts.inter(
                        fontSize: 11,
                        color: tc.textMuted,
                        fontWeight: FontWeight.w400,
                      ),
                    ),
                  ),
                ),
              ),

            const SizedBox(height: AppSpacing.s24),

            // ── Status text ──
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: QiblaLayout.screenPadding),
              child: Center(
                child: facingQibla
                    ? Text(
                        statusText,
                        style: GoogleFonts.inter(
                          fontSize: QiblaLayout.statusFontSize,
                          color: tc.accent,
                          fontWeight: FontWeight.w600,
                        ),
                      )
                    : (_compassAvailable && _heading != null)
                        ? Text(
                            statusText,
                            style: GoogleFonts.inter(
                              fontSize: QiblaLayout.statusFontSize,
                              color: tc.textMuted,
                              fontWeight: FontWeight.w400,
                            ),
                          )
                        : RichText(
                            textAlign: TextAlign.center,
                            text: TextSpan(
                              style: GoogleFonts.inter(
                                fontSize: QiblaLayout.statusFontSize,
                                color: tc.textMuted,
                                fontWeight: FontWeight.w400,
                              ),
                              children: [
                                const TextSpan(text: 'Qibla is at '),
                                TextSpan(
                                  text: '${qiblaBearing.toStringAsFixed(1)}°',
                                  style: TextStyle(
                                    color: tc.accent,
                                    fontWeight: FontWeight.w600,
                                  ),
                                ),
                                const TextSpan(text: ' from North.'),
                              ],
                            ),
                          ),
              ),
            ),
            const SizedBox(height: AppSpacing.s24),

            // ── Kaaba label card ──
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: QiblaLayout.screenPadding),
              child: Container(
                padding: const EdgeInsets.all(QiblaLayout.cardPadding),
                decoration: BoxDecoration(
                  color: tc.card,
                  borderRadius: BorderRadius.circular(QiblaLayout.cardRadius),
                  border: Border.all(color: tc.cardBorder, width: 1),
                ),
                child: Column(
                  children: [
                    Text(
                      'الكعبة المشرّفة',
                      style: GoogleFonts.inter(
                        fontSize: QiblaLayout.arabicFontSize,
                        fontWeight: FontWeight.w600,
                        color: tc.accent,
                      ),
                      textDirection: TextDirection.rtl,
                    ),
                    const SizedBox(height: 4),
                    Text(
                      'Al-Kaaba Al-Musharrafah',
                      style: GoogleFonts.inter(
                        fontSize: QiblaLayout.translitFontSize,
                        fontWeight: FontWeight.w500,
                        color: tc.textMuted,
                      ),
                    ),
                  ],
                ),
              ),
            ),
            const SizedBox(height: QiblaLayout.screenPadding),
          ],
        );
      },
    );
  }
}

// ────────────────────────────────────────────────
// POINTER PAINTER (fixed triangle at top, does NOT rotate)
// ────────────────────────────────────────────────
class _PointerPainter extends CustomPainter {
  final Color color;
  _PointerPainter({required this.color});

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()..color = color;
    final path = Path()
      ..moveTo(size.width / 2, size.height)
      ..lineTo(0, 0)
      ..lineTo(size.width, 0)
      ..close();
    canvas.drawPath(path, paint);
  }

  @override
  bool shouldRepaint(_PointerPainter old) => old.color != color;
}

// ────────────────────────────────────────────────
// COMPASS PAINTER (rotates with device heading)
// ────────────────────────────────────────────────
class _CompassPainter extends CustomPainter {
  final double qiblaDegrees;
  final Color accent;
  final Color muted;
  final Color ring;
  final Color needle;

  _CompassPainter({
    required this.qiblaDegrees,
    required this.accent,
    required this.muted,
    required this.ring,
    required this.needle,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final center = Offset(size.width / 2, size.height / 2);
    final radius = size.width / 2 - 20;

    // ── Outer ring ──
    final ringPaint = Paint()
      ..color = ring
      ..style = PaintingStyle.stroke
      ..strokeWidth = QiblaLayout.compassStroke;
    canvas.drawCircle(center, radius, ringPaint);

    // ── Tick marks ──
    final tickPaint = Paint()
      ..color = muted.withValues(alpha: 0.5)
      ..strokeWidth = 1.0;
    final majorTickPaint = Paint()
      ..color = muted.withValues(alpha: 0.8)
      ..strokeWidth = 1.5;

    for (int deg = 0; deg < 360; deg += 5) {
      final isMajor = deg % 90 == 0;
      final isMinor30 = deg % 30 == 0;
      final len = isMajor
          ? QiblaLayout.tickLengthMajor
          : (isMinor30 ? QiblaLayout.tickLength + 2 : QiblaLayout.tickLength);
      final rad = deg * pi / 180 - pi / 2;
      final outerX = center.dx + radius * cos(rad);
      final outerY = center.dy + radius * sin(rad);
      final innerX = center.dx + (radius - len) * cos(rad);
      final innerY = center.dy + (radius - len) * sin(rad);
      canvas.drawLine(
        Offset(outerX, outerY),
        Offset(innerX, innerY),
        isMajor || isMinor30 ? majorTickPaint : tickPaint,
      );
    }

    // ── Cardinal labels ──
    final cardinals = {'N': 0, 'E': 90, 'S': 180, 'W': 270};
    cardinals.forEach((label, deg) {
      final rad = deg * pi / 180 - pi / 2;
      final labelRadius = radius - QiblaLayout.tickLengthMajor - 12;
      final x = center.dx + labelRadius * cos(rad);
      final y = center.dy + labelRadius * sin(rad);
      final isNorth = label == 'N';
      final tp = TextPainter(
        text: TextSpan(
          text: label,
          style: TextStyle(
            fontSize: QiblaLayout.cardinalFontSize,
            fontWeight: FontWeight.w700,
            color: isNorth ? accent : muted.withValues(alpha: 0.6),
          ),
        ),
        textDirection: TextDirection.ltr,
      )..layout();
      tp.paint(canvas, Offset(x - tp.width / 2, y - tp.height / 2));
    });

    // ── Qibla needle ──
    final needleRad = qiblaDegrees * pi / 180 - pi / 2;
    final needlePaint = Paint()
      ..color = accent
      ..strokeWidth = QiblaLayout.needleWidth
      ..strokeCap = StrokeCap.round;
    final needleEnd = Offset(
      center.dx + (radius - QiblaLayout.tickLengthMajor - 24) * cos(needleRad),
      center.dy + (radius - QiblaLayout.tickLengthMajor - 24) * sin(needleRad),
    );
    canvas.drawLine(center, needleEnd, needlePaint);

    // ── Center dot ──
    canvas.drawCircle(center, QiblaLayout.centerDotRadius, Paint()..color = accent);

    // ── Kaaba marker on ring ──
    final kaabaX = center.dx + (radius + 2) * cos(needleRad);
    final kaabaY = center.dy + (radius + 2) * sin(needleRad);
    canvas.drawCircle(Offset(kaabaX, kaabaY), QiblaLayout.kaabaIconSize / 2 + 3, Paint()..color = accent);
    final kaabaRect = Rect.fromCenter(
      center: Offset(kaabaX, kaabaY),
      width: QiblaLayout.kaabaIconSize * 0.55,
      height: QiblaLayout.kaabaIconSize * 0.55,
    );
    canvas.drawRect(kaabaRect, Paint()..color = Colors.black..style = PaintingStyle.fill);
  }

  @override
  bool shouldRepaint(_CompassPainter old) =>
      old.qiblaDegrees != qiblaDegrees || old.accent != accent;
}
